src/service/src/grpc.rs

```diff
@@ -477,40 +479,86 @@ impl Interceptor for VersionAttachInterceptor {
     fn call(&mut self, mut request: Request<()>) -> Result<Request<()>, Status> {
         request
             .metadata_mut()
-            .insert(VERSION_METADATA_KEY.clone(), self.version.clone());
+            .insert(VERSION_HEADER_KEY, self.version.clone());
         Ok(request)
     }
 }
 
-/// A gRPC interceptor that ensures the version attached to the request by the
-/// `VersionAttachInterceptor` exactly matches the expected version.
-#[derive(Debug, Clone)]
-struct VersionCheckExactInterceptor {
-    version: AsciiMetadataValue,
+/// A `tower` layer that validates requests for compatibility with the server.
+#[derive(Clone)]
+struct RequestValidationLayer {
+    version: Version,
+    process_name: Option<String>,
```

Comment:
An interceptor doesn't give access to the request URI, and AFAICT there is no other way to get at the hostname, so I had to implement a `tower` layer instead. Decided to merge the existing version check into that to avoid some redundancy.